execution/order_by.go

```diff
@@ -18,132 +15,217 @@ const (
 	}
 
-	defer func() {
-		if err := recover(); err != nil {
-			stream = nil
-			outErr = errors.Wrap(err.(error), "couldn't sort records")
+	keyPrefix := append(append([]byte("$"), key.MonotonicMarshal()...), '$')
+	txByKey := tx.WithPrefix(keyPrefix)
+
+	var recordValueMap *storage.Map
+	var recordCountMap *storage.Map
+
+
+	recordValueMap = storage.NewMap(txByKey.WithPrefix(recordValuePrefix))
+	recordCountMap = storage.NewMap(txByKey.WithPrefix(recordCountPrefix))
+
+	mapping := make(map[string]octosql.Value, len(record.Fields()))
+	for _, field := range record.Fields() {
+		mapping[field.Name.String()] = record.Value(field.Name)
+	}
+	variables := record.AsVariables()
+
+
+	recordPrefix := []byte("$")
+
+	for i := range ob.expressions {
+		expressionValue, err := ob.expressions[i].ExpressionValue(ctx, variables)
+		if err != nil {
+			return errors.Wrap(err, "couldn't evaluate expression")
 		}
-	}()
-	sort.Slice(records, func(i, j int) bool {
-		iRec := records[i]
-		jRec := records[j]
-
-		for num, expr := range expressions {
-			// TODO: Aggressive caching of these expressions...
-			iVars, err := variables.MergeWith(iRec.AsVariables())
-			if err != nil {
-				panic(errors.Wrap(err, "couldn't merge variables"))
-			}
-			jVars, err := variables.MergeWith(jRec.AsVariables())
+		if ob.directions[i] == Ascending {
+			recordPrefix = append(recordPrefix, expressionValue.MonotonicMarshal()...)
+		} else {
+			recordPrefix = append(recordPrefix, expressionValue.ReversedMonotonicMarshal()...)
+		}
+		recordPrefix = append(recordPrefix, '$')
+	}
+
+	bytePref := append(append(recordPrefix, []byte(NewRecordFromRecord(record, WithNoUndo()).String())...), '$')
+	pref := OrderByKey{key: bytePref}
+
+	count  := octosql.MakePhantom()
+	err := recordCountMap.Get(&pref, &count)
+
+	if !record.IsUndo() {
+		if err == storage.ErrNotFound {
+
+			err = recordValueMap.Set(&pref, record)
 			if err != nil {
-				panic(errors.Wrap(err, "couldn't merge variables"))
+				return err
 			}
 
-			x, err := expr.ExpressionValue(ctx, iVars)
+			count = octosql.MakeInt(1)
+			err = recordCountMap.Set(&pref, &count)
 			if err != nil {
-				panic(errors.Wrapf(err, "couldn't get order by expression with index %v value", num))
+				return err
 			}
-			y, err := expr.ExpressionValue(ctx, jVars)
+		} else if err != nil {
+			return err
+		} else {
+			// Keep track of record vs retraction count
+			count = octosql.MakeInt(count.AsInt() + 1)
+			err = recordCountMap.Set(&pref, &count)
 			if err != nil {
-				panic(errors.Wrapf(err, "couldn't get order by expression with index %v value", num))
+				return err
 			}
-
-			if !isSorteable(x) {
-				panic(errors.Errorf("value %v of type %v is not comparable", x, reflect.TypeOf(x).String()))
-			}
-			if !isSorteable(y) {
-				panic(errors.Errorf("value %v of type %v is not comparable", y, reflect.TypeOf(y).String()))
+		}
+	} else {
+		if err == storage.ErrNotFound {
+			return errors.Wrap(err, "couldn't retract record")
```

Comment:
This is a tricky case, but You should handle early retractions here, which means save count -1 instead of returning error